File: src/renderer/services/backup/formats/version3.ts
import { z } from 'zod'
import { DocHeader } from './version0'
import { Settings } from './version2'
import type { Import } from './version4'

/**
 * Layout v3, will be used in Export v3.
 */
export type Layouts = z.output<typeof Layouts>
export const Layouts = z.object({
  sources: z.array(
    DocHeader.extend({
      order: z.number().nonnegative().finite().default(0),
      title: z.string().min(1),
      image: z.string().min(1).nullable()
    })
  ),
  devices: z.array(
    DocHeader.extend({
      driverId: z.string().uuid(),
      title: z.string(),
      path: z.string()
    })
  ),
  ties: z.array(
    DocHeader.extend({
      sourceId: z.string().uuid(),
      deviceId: z.string().uuid(),
      inputChannel: z.number().int(),
      outputChannels: z
        .object({
          video: z.number().int().optional(),
          audio: z.number().int().optional()
        })
        .default({})
    })
  )
})

/**
 * Backup parser for v3.
 */
export const Backup = z
  .object({
    version: z.literal(3),
    settings: Settings,
    layouts: Layouts
  })
  .transform(function transformToV4(payload): Import {
    const sortedSources = payload.layouts.sources.sort((a, b) => a.order - b.order)
    const buttonOrder = sortedSources.map((source) => source._id)

    return {
      version: 'normalized' as const,
      settings: { ...payload.settings, buttonOrder },
      layouts: {
        sources: payload.layouts.sources,
        devices: payload.layouts.devices,
        ties: payload.layouts.ties.map(({ outputChannels, ...tie }) => ({
          ...tie,
          outputChannels: {
            video: outputChannels.video ?? undefined,
            audio: outputChannels.audio ?? undefined
          }
        }))
      }
    }
  })

File: composeApp/src/jvmMain/kotlin/org/sleepingcats/bridgecmdr/common/backup/model/Version3.kt
@file:OptIn(ExperimentalUuidApi::class)

package org.sleepingcats.bridgecmdr.common.backup.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.sleepingcats.bridgecmdr.common.setting.IconSize
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

object Version3 : BackupParser<Version5.Backup> {
  @Serializable
  data class Source(
    @SerialName("_id")
    val id: Uuid,
    val order: Int = 0,
    val title: String,
    val image: String? = null,
  )

  @Serializable
  data class Device(
    @SerialName("_id")
    val id: Uuid,
    val driverId: Uuid,
    val title: String,
    val path: String,
  )

  @Serializable
  data class Tie(
    @SerialName("_id")
    val id: Uuid,
    val sourceId: Uuid,
    val deviceId: Uuid,
    val inputChannel: Int,
    val outputChannels: OutputChannels = OutputChannels(),
  ) {
    @Serializable
    data class OutputChannels(
      val video: Int? = null,
      val audio: Int? = null,
    )
  }

  @Serializable
  data class Layouts(
    val sources: List<Source>,
    val devices: List<Device>,
    val ties: List<Tie>,
  )

  @Serializable
  data class Backup(
    val version: Int,
    val settings: Version2.Settings = Version2.Settings(),
    val layouts: Layouts,
  )

  fun upgradeBackup(backup: Backup) =
    Version5.Backup(
      version = 5,
      settings =
        Version5.Settings(
          appTheme = backup.settings.colorScheme.newValue,
          iconSize = IconSize.entries.find { entry -> entry.size == backup.settings.iconSize } ?: IconSize.Normal,
          powerOnDevicesAtStart = backup.settings.powerOnSwitchesAtStart,
          powerOffTaps = backup.settings.powerOffWhen.newValue,
          buttonOrder =
            backup.layouts.sources
              .sortedBy { it.order }
              .map { it.id },
        ),
      layouts =
        Version4.Layouts(
          sources =
            backup.layouts.sources.map {
              Version4.Source(
                id = it.id,
                title = it.title,
                image = it.image,
              )
            },
          devices =
            backup.layouts.devices.map {
              Version4.Device(
                id = it.id,
                driverId = it.driverId,
                title = it.title,
                path = it.path,
              )
            },
          ties =
            backup.layouts.ties.map {
              Version4.Tie(
                id = it.id,
                sourceId = it.sourceId,
                deviceId = it.deviceId,
                inputChannel = it.inputChannel,
                outputVideoChannel = it.outputChannels.video,
                outputAudioChannel = it.outputChannels.audio,
              )
            },
        ),
    )

  override fun parseBackup(raw: String) =
    upgradeBackup(
      Json.decodeFromString<Backup>(raw).also {
        check(it.version == 3) { "Wrong backup version, expected 3" }
      },
    )
}
